src/main/java/com/snowflake/kafka/connector/internal/ingestsdk/ClientManager.java

```diff
@@ -0,0 +1,122 @@
+    Map<String, String> streamingPropertiesMap =
+        StreamingUtils.convertConfigForStreamingClient(new HashMap<>(connectorConfig));
+    Properties streamingClientProps = new Properties();
+    streamingClientProps.putAll(streamingPropertiesMap);
+
+    String streamingIngestClientName = this.getStreamingIngestClientName(kcInstanceId);
+
+    try {
+      LOGGER.info("Creating Streaming Client. ClientName:{}", streamingIngestClientName);
+      this.streamingIngestClientCount++;
+      this.streamingIngestClient =
+          SnowflakeStreamingIngestClientFactory.builder(streamingIngestClientName)
+              .setProperties(streamingClientProps)
+              .build();
+    } catch (SFException ex) {
+      LOGGER.error(
+          "Exception creating streamingIngestClient with name:{}", streamingIngestClientName);
+      throw new ConnectException(ex);
+    }
+  }
+
+  /**
+   * Calls the ingest sdk to close the client sdk
+   *
+   * @return true if the client was successfully closed, false if not
+   */
+  public boolean closeStreamingClient() {
+    if (this.streamingIngestClient == null || this.streamingIngestClient.isClosed()) {
+      LOGGER.info("Streaming client is already closed or null");
+      return true;
+    }
+
+    String streamingIngestClientName = this.streamingIngestClient.getName();
+    LOGGER.info("Closing Streaming Client:{}", streamingIngestClientName);
+
+    try {
+      this.streamingIngestClient.close();
+      return true;
+    } catch (Exception e) {
+      String message =
+          e.getMessage() != null && !e.getMessage().isEmpty()
+              ? e.getMessage()
+              : "no error message provided";
+
+      String cause =
+          e.getCause() != null
+                  && e.getCause().getStackTrace() != null
+                  && !Arrays.toString(e.getCause().getStackTrace()).isEmpty()
+              ? Arrays.toString(e.getCause().getStackTrace())
+              : "no cause provided";
+
+      LOGGER.error("Failure closing Streaming client msg:{}, cause:{}", message, cause);
+      return false;
+    }
+  }
+
+  /**
+   * Gets the streaming client if it was created
+   *
+   * @return The streaming client, throws an exception if no client was initialized
+   */
+  public SnowflakeStreamingIngestClient getStreamingIngestClient() {
+    if (this.streamingIngestClient != null && !this.streamingIngestClient.isClosed()) {
+      return this.streamingIngestClient;
+    }
+
+    LOGGER.error("Streaming ingest client was null or closed. It must be initialized");
+    throw SnowflakeErrors.ERROR_3009.getException();
+  }
+
+  /**
+   * Gets the clients name by adding a prefix and client count
+   *
+   * @param kcInstanceId the indentifier for the connector creating this client
+   * @return the streaming ingest client name
+   */
+  private String getStreamingIngestClientName(String kcInstanceId) {
+    return STREAMING_CLIENT_PREFIX_NAME + kcInstanceId + this.streamingIngestClientCount;
+  }
```

Comment:
Is it possible to call this twice and get back the same name for two separate clients anywhere?

Reply:
With the new method - buildStreamingIngestClientName(String kcInstanceId, int clientId) - yes you could. There is no streamingIngestClient id so we use the client name to reference clients. I chose to not use a permanently increasing id because it may introduce race conditions - wanted to keep state out of the KcStreamingIngestClient, but happy to discuss this one further